enhancements/20250203-third-party-storage-support.md

```diff
@@ -0,0 +1,98 @@
+
+## Summary
+
+This enhancement aims to provide a generic way to leverage third-party storage solutions which already have CSI driver. By introducing Containerized-Data Importer (CDI), Harvester can support third-party storage solutions with the current Virtual Machine Image path for the root and data disks.
+
+### Related Issues
+
+https://github.com/harvester/harvester/issues/1199
+
+## Motivation
+
+### Goals
+
+To provide a generic way to leverage third-party storage solutions with the current Virtual Machine Image path for root and data disks.
+
+### Non-goals [optional]
+
+Virtual Machine Backup and Restore with third-party storage are not in the scope of this enhancement.
+
+## Proposal
+
+We will introduce Containerized-Data Importer (CDI) to abstract the Virtual Machine Image path for generic third-party storage solutions.
+The CDI will respond to the VM Image and the root disk with the third-party storage solution.
+
+### User Stories
+
+#### User wants to use LVM for better performance and latency 
+
+Harvester introduces the LVM CSI driver v1.4.0. But the user can now only use it for the data disk. The user wants to use LVM for both the root and data disks.
+
+#### Users already have a storage solution on their data center and wants to use it in Harvester
+
+Most users already have a storage solution in their data center (e.g. Ceph, HPE, NetApp, etc.). They want to use their existing storage solution in Harvester for data and root disks.
+
+### User Experience In Detail
+
+This enhancement will abstract the Virtual Machine Image path generically. So, the user only needs to choose the corresponding StorageClass provided by the third-party storage solution. They can create a VM Image with their storage solution. Then, they can create a VM with the VM Image.
+
+### API changes
+
+TBD
+
+## Design
+
+### High-level Architecture
+
+![3rd-party-high-level](./20250203-third-party-storage-support/high-level-design.png)
+
+### Implementation Overview
+
+We could split the implementation into the following parts:
+- Introduce Containerized Data Importer (CDI) to Harvester
+- VM Image Creation through CDI
+- Root Disk Creation through CDI
+
+#### Introduce Containerized Data Importer (CDI) to Harvester
+
+To introduce Containerized-Data Importer (CDI) to Harvester, we need to add the charts and the CRDs to Harvester. The CDI will be responsible for the VM Image and the root disk creation with the third-party storage solution. 
+
+#### VM Image Creation through CDI
+
+The VM Image creation design could be like the following:
+![vm-image-design](./20250203-third-party-storage-support/vm-image-design.png)
+
+Because the target PVC (golden image) size should be defined before DataVolume creation, we need to prefetch the size and virtual size.
+We have different mechanisms for different source types (download/upload) for size/virtual size prefetching, and progress updates.
```

Comment:
Will download be done by CDI too? What are the supported download targets (ftp, http, oci etc.)?

Reply:
Will we be supporting both [sync](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/upload.md#synchronous) and [async](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/upload.md#asynchronous) upload?

Reply:
We now only support the download from http and sync upload.

CDI has better capability, but we are trying to align it with our previous version.
Maybe we could try to improve that in a later version.